core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -157,22 +164,29 @@ private class ShuffleStatus(
       invalidateSerializedMapOutputStatusCache()
     }
     mapStatuses(mapIndex) = status
+    mapIdToMapIndex(status.mapId) = mapIndex
   }
 
   /**
    * Update the map output location (e.g. during migration).
    */
   def updateMapOutput(mapId: Long, bmAddress: BlockManagerId): Unit = withWriteLock {
     try {
-      val mapStatusOpt = mapStatuses.find(x => x != null && x.mapId == mapId)
+      // OpenHashMap would return 0 if the key doesn't exist.
+      val mapIndex = if (mapIdToMapIndex.contains(mapId)) {
+        Some(mapIdToMapIndex(mapId))
+      } else {
+        None
+      }
```

Comment:
Thoughts on changing this to either:
* Add a `get(key): Option[V]` to `OpenHashMap` or
* Simply do `Option(mapIdToMapIndex(mapId))` ?

The second option relies on implementation detail that we return `null` when `apply` does not find a `key` - while the first option is more principled change.

Reply:
The problem is OpenHashMap casts null value to the declared type internally. Thus if the key doesn't exist, it would cast `null` to Int, and the output would be 0. It is not possible for us tell whether the key doesn't exist or the value is really 0, without calling the `contains` function.

I'll try to follow the first option to add a `get(key): Option[V]`.